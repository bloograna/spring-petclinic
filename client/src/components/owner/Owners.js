import React from 'react';
import PropTypes from 'prop-types';
import { Table, Button } from 'react-bootstrap';
import { isEmpty, capitalize } from 'lodash';

const Owners = ({
  owners,
  pets,
  onAddPet,
  setActivePet,
  getPetsByOwnerId,
  getVisitByPetId
}) =>
  isEmpty(owners) ? null : (
    <Table striped bordered hover variant="dark">
      {constructTableHeader()}
      <tbody>
        {constructTableRows(
          owners,
          pets,
          onAddPet,
          setActivePet,
          getPetsByOwnerId,
          getVisitByPetId
        )}
      </tbody>
    </Table>
  );

const constructTableRows = (
  owners,
  pets,
  onAdd,
  setActivePet,
  onLookUp,
  getVisitByPetId
) => {
  const rows = [];
  owners.forEach(owner => {
    if (owner) {
      const petButtons = contructPetButtons(
        owner,
        pets,
        setActivePet,
        onLookUp,
        getVisitByPetId
      );
      rows.push(
        constructTableRow(
          owner.id,
          owner.firstName,
          owner.lastName,
          petButtons,
          onAdd
        )
      );
    }
  });
  return rows;
};

const attachOwnerId = (ownerId, action) => {
  action(ownerId);
};

const contructPetButtons = (
  owner,
  pets,
  setActivePet,
  onLookUp,
  getVisitByPetId
) => {
  return pets && pets[owner.id] ? (
    Object.values(pets[owner.id]).map(pet => {
      const petName = capitalize(pet.name);
      return (
        <Button
          variant="outline-light"
          size="sm"
          key={`pet-info-${pet.id}`}
          onClick={() => {
            setActivePet(pet);
            getVisitByPetId(pet.id);
          }}
        >
          {petName}
        </Button>
      );
    })
  ) : (
    <Button
      variant="outline-info"
      size="sm"
      key="pet-info-search"
      onClick={() => attachOwnerId(owner.id, onLookUp)}
    >
      View Pets
    </Button>
  );
};

const constructTableRow = (id, firstName, lastName, pets, onAdd) => (
  <tr key={`owner-table-row-${id}`}>
    <td>{id}</td>
    <td>{firstName}</td>
    <td>{lastName}</td>
    <td>
      {pets}
      <Button variant="link" size="sm" onClick={() => attachOwnerId(id, onAdd)}>
        Add
      </Button>
    </td>
  </tr>
);

const constructTableHeader = () => (
  <thead>
    <tr>
      <th>#</th>
      <th>First Name</th>
      <th>Last Name</th>
      <th>Pets</th>
    </tr>
  </thead>
);

Owners.propTypes = {
  owners: PropTypes.array.isRequired,
  pets: PropTypes.array.isRequired,
  onAddPet: PropTypes.func.isRequired,
  setActivePet: PropTypes.func.isRequired,
  getPetsByOwnerId: PropTypes.func.isRequired,
  getVisitByPetId: PropTypes.func.isRequired
};

export default Owners;
