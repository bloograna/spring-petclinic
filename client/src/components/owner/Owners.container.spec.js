import React from 'react';
import { TestOwnersContainer } from './Owners.container';

describe('OwnersContainer render', () => {
  let props;
  let wrapper;
  let instance;
  let event;

  beforeEach(() => {
    props = {
      dispatch: action => action,
      searchResults: [],
      showAddOwnerModal: false,
      shouldValidateOwnerModalData: false,
      pets: [],
      petTypes: [],
      showAddPetModal: false,
      shouldValidatePetModalData: false,
      activePet: null,
      // actions
      searchByLastName: jest.fn(),
      openAddOwnerModal: jest.fn(),
      closeAddOwnerModal: jest.fn(),
      validateOwnerModalData: jest.fn(),
      saveOwner: jest.fn(),
      openAddPetModal: jest.fn(),
      closeAddPetModal: jest.fn(),
      setActivePet: jest.fn(),
      getPetsByOwner: jest.fn(),
      validatePetModalData: jest.fn(),
      savePet: jest.fn()
    };

    event = {
      target: {
        value: 'value'
      },
      preventDefault: jest.fn()
    };

    wrapper = shallow(<TestOwnersContainer {...props} />);
    instance = wrapper.instance();
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders successfully', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('onSearch', () => {
    const { searchByLastName } = props;
    const lastName = 'lastName';
    const localEvent = {
      currentTarget: {
        form: [{ value: lastName }]
      }
    };

    instance.onSearch(localEvent);
    expect(searchByLastName).toHaveBeenCalledWith(lastName);
  });

  test('handleAddPetFormData should save when valid', () => {
    const { validatePetModalData, savePet } = props;
    const checkValidity = jest.fn();
    const localEvent = {
      currentTarget: {
        checkValidity: checkValidity
      }
    };

    checkValidity.mockReturnValue(true);

    const formObj = { value: 'foo' };

    instance.handleAddPetFormData(formObj, localEvent);
    expect(validatePetModalData).toHaveBeenCalled();
    expect(savePet).toHaveBeenCalledWith(formObj);
  });

  test('handleAddPetFormData should not save when invalid', () => {
    const { validatePetModalData, savePet } = props;
    const checkValidity = jest.fn();
    const localEvent = {
      currentTarget: {
        checkValidity: checkValidity
      }
    };

    checkValidity.mockReturnValue(false);

    const formObj = { value: 'foo' };

    instance.handleAddPetFormData(formObj, localEvent);
    expect(validatePetModalData).toHaveBeenCalled();
    expect(savePet).not.toHaveBeenCalled();
  });
});
