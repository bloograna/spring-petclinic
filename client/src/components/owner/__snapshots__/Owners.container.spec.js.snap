// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnersContainer render renders successfully 1`] = `
<div>
  <AddOwnerModal
    onAddButtonClick={[Function]}
    onHideAddOwnerModal={[MockFunction]}
    shouldValidateOwnerModalData={false}
    showAddOwnerModal={false}
  />
  <AddPetModal
    onAddButtonClick={[Function]}
    onHideAddPetModal={[MockFunction]}
    pet={null}
    petTypes={Array []}
    shouldValidatePetModalData={false}
    showAddPetModal={false}
    visits={Array []}
  />
  <ForwardRef(Bootstrap(Form))
    noValidate={true}
  >
    <ForwardRef(Bootstrap(FormRow))>
      <ForwardRef(Bootstrap(FormGroup))
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          }
        }
        controlId="lastName"
        md="4"
      >
        <ForwardRef(ContextTransform)>
          Last Name
        </ForwardRef(ContextTransform)>
        <ForwardRef(ContextTransform)
          placeholder="Last Name"
          type="text"
        />
        <ForwardRef(Bootstrap(Button))
          onClick={[Function]}
          variant="primary"
        >
          Search
        </ForwardRef(Bootstrap(Button))>
      </ForwardRef(Bootstrap(FormGroup))>
    </ForwardRef(Bootstrap(FormRow))>
    <ForwardRef(Bootstrap(Button))
      onClick={[MockFunction]}
      variant="primary"
    >
      Add an owner
    </ForwardRef(Bootstrap(Button))>
  </ForwardRef(Bootstrap(Form))>
  <Owners
    getPetsByOwnerId={[MockFunction]}
    onAddPet={[MockFunction]}
    owners={Array []}
    pets={Array []}
    setActivePet={[Function]}
  />
</div>
`;
